### PR TITLE
RELATED: SD-936 add bootstrap module to fetch all data from resource

### PR DESCRIPTION
**Pull request by dlcd2203**
---

**Proposed release type:** minor

# PR checklist

- [x] Change was tested using dev-release in Analytical Designer and Dashboards (if applicable).
- [x] Change is described in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Migration guide (for major update) is written to [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] The proposed release type is appropriate (see the comment in [PR template](../blob/master/.github/pull_request_template.md))

Related ticket: https://jira.intgdc.com/browse/SD-936
GD-Labs: https://github.com/gooddata/gd-labs/pull/397
AD: https://github.com/gooddata/gdc-analytical-designer/pull/3063
KD: https://github.com/gooddata/gdc-dashboards/pull/2781

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/101/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/537/ : SUCCESS in 1m 19s

File: src/project.ts [hunk 1]
```diff
@@ -82,15 +82,22 @@ export class ProjectModule {
...
+     */
+    public getCurrentProjectIdInBootstrap(bootstrapData: any) {
+        const currentProject = bootstrapData.bootstrapResource.current.project;
+        // handle situation in which current project is missing (e.g. new user)
+        if (!currentProject) {
+            return null;
+        }
+
+        return bootstrapData.bootstrapResource.current.project.links.self.split("/").pop();
     }
```
Comment (ptquang86) [inline src/project.ts:101]: For consistency, please consider to 
- Move thin function into `src/bootstrap.ts` or 
- Move `getMapboxToken` out of `src/bootstrap.ts`

They have the same functionality. 
I prefer the second option :)

File: src/user.ts [hunk 2]
```diff
@@ -177,18 +177,28 @@ export class UserModule {
...
         });
     }
 
+    /**
+     * Returns current user info from bootstrapData
+     * @method getAccountInfoInBootstrap
+     * @param bootstrapData - data was got from bootstrap resource
+     */
+    public getAccountInfoInBootstrap(bootstrapData: any) {
+        const { bootstrapResource } = bootstrapData;
```
Comment (ptquang86) [inline src/user.ts:191]: How about
`const { bootstrapResource: { accountSetting: { ... }, current: { ... }, settings: { ... } } } = bootstrapData;`?

File: test/bootstrap.test.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                            loginMD5: "LOGIN_MD5",
+                            mapboxToken: "mapbox token",
+                        },
+                        settings: {
+                            organizationName: "ORG_NAME",
+                        },
+                    },
+                });
+            });
+    });
```
Comment (naibinh) [inline test/bootstrap.test.ts:64]: Please add one case for mapbox token

File: src/project.ts [hunk 4]
```diff
@@ -82,15 +82,22 @@ export class ProjectModule {
...
+    }
 
-                return result.bootstrapResource.current.project.links.self.split("/").pop();
-            });
+    /**
+     * Return current project id in bootstrap
+     * @method getCurrentProjectIdInBootstrap
+     * @param bootstrapData - data was got from bootstrap resource
+     */
+    public getCurrentProjectIdInBootstrap(bootstrapData: any) {
```
[the remark this answers is not in the document]
Comment (dlcd2203) [reply src/project.ts]: fixed

File: test/bootstrap.test.ts (hunk 3 above)
Comment (dlcd2203) [reply test/bootstrap.test.ts:58]: fixed

File: src/user.ts (hunk 2 above)
Comment (dlcd2203) [reply src/user.ts]: fixed

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/102/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/538/ : SUCCESS in 1m 12s

File: src/user.ts [hunk 5]
```diff
@@ -188,14 +188,25 @@ export class UserModule {
...
      */
     public getAccountInfoInBootstrap(bootstrapData: any) {
-        const { bootstrapResource } = bootstrapData;
+        const {
+            bootstrapResource: {
+                accountSetting: {
+                    login,
+                    firstName,
+                    lastName,
+                    links: { self },
```
Comment (naibinh) [inline src/user.ts:197]: We can use alias as `links: { self: profileUri },`

Comment (dlcd2203) [reply src/user.ts]: fixed

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/103/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/539/ : SUCCESS in 53s

**Comment by naibinh** [comment]
LGTM

**Comment by ptquang86** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-oss-check/106/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/542/ : SUCCESS in 1m 20s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-js-merge-compliance-check/124/ : SUCCESS in 13s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-js-unit-tests-zuul-docker-el7/543/ : SUCCESS in 1m 22s

**Pull request merged**